documents/extension/funding/service.go

```diff
@@ -0,0 +1,277 @@
+package funding
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	"github.com/centrifuge/go-centrifuge/utils"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/centrifuge/go-centrifuge/centerrors"
+	"github.com/centrifuge/go-centrifuge/documents"
+	"github.com/centrifuge/go-centrifuge/errors"
+	clientfundingpb "github.com/centrifuge/go-centrifuge/protobufs/gen/go/funding"
+)
+
+// Service defines specific functions for extension funding
+type Service interface {
+	documents.Service
+	// DeriveFromPayload derives Funding from clientPayload
+	DeriveFromPayload(ctx context.Context, req *clientfundingpb.FundingCreatePayload, identifier []byte) (documents.Model, error)
+
+	// DeriveFundingResponse returns a funding in client format
+	DeriveFundingResponse(model documents.Model, fundingID string) (*clientfundingpb.FundingResponse, error)
+}
+
+// service implements Service and handles all funding related persistence and validations
+type service struct {
+	documents.Service
+	tokenRegFinder func() documents.TokenRegistry
+}
+
+const (
+	fundingLabel    = "centrifuge_funding"
+	fundingFieldKey = "centrifuge_funding[{IDX}]."
+	fundingIDx      = "{IDX}"
+	fundingIDLabel  = "funding_id"
+)
+
+// DefaultService returns the default implementation of the service.
+func DefaultService(
+	srv documents.Service,
+	tokenRegFinder func() documents.TokenRegistry,
+) Service {
+	return service{
+		Service:        srv,
+		tokenRegFinder: tokenRegFinder,
+	}
+}
+
+func newFundingID() string {
+	return hexutil.Encode(utils.RandomSlice(32))
+}
+
+func generateLabel(idx, fieldName string) string {
+	return strings.Replace(fundingFieldKey, fundingIDx, idx, -1) + fieldName
+}
+
+func keyFromJSONTag(idx, jsonTag string) (key string, err error) {
+	correctJSONParts := 2
+	jsonKeyIdx := 0
+
+	// example `json:"days,omitempty"`
+	jsonParts := strings.Split(jsonTag, ",")
+	if len(jsonParts) == correctJSONParts {
+		return generateLabel(idx, jsonParts[jsonKeyIdx]), nil
+
+	}
+	return key, ErrNoFundingField
+
+}
+
+func getFundingsLatestIdx(model documents.Model) (idx *documents.Int256, err error) {
+	key, err := documents.AttrKeyFromLabel(fundingLabel)
+	if err != nil {
+		return idx, err
+	}
+
+	attr, err := model.GetAttribute(key)
+	if err != nil {
+		return idx, err
+	}
+
+	idx = attr.Value.Int256
+
+	z, err := documents.NewInt256("0")
+	if err != nil {
+		return idx, err
```

Comment:
can we return here `nil` instead as the idx ?

Reply:
We could if we use a string pointer

Reply:
I meant that as idx is nil until we assign: `idx = attr.Value.Int256` then after any error will have the value of the attr. to be consistent is better to return `nil` here explicitely. Unless I am missing something else.